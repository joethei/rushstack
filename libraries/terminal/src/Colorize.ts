// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { AnsiEscape } from './AnsiEscape';

export enum SgrParameterAttribute {
  BlackForeground = 30,
  RedForeground = 31,
  GreenForeground = 32,
  YellowForeground = 33,
  BlueForeground = 34,
  MagentaForeground = 35,
  CyanForeground = 36,
  WhiteForeground = 37,
  GrayForeground = 90,
  DefaultForeground = 39,

  BlackBackground = 40,
  RedBackground = 41,
  GreenBackground = 42,
  YellowBackground = 43,
  BlueBackground = 44,
  MagentaBackground = 45,
  CyanBackground = 46,
  WhiteBackground = 47,
  GrayBackground = 100,
  DefaultBackground = 49,

  Bold = 1,

  // On Linux, the "BoldOff" code instead causes the text to be double-underlined:
  // https://en.wikipedia.org/wiki/Talk:ANSI_escape_code#SGR_21%E2%80%94%60Bold_off%60_not_widely_supported
  // Use "NormalColorOrIntensity" instead
  // BoldOff = 21,

  Dim = 2,
  NormalColorOrIntensity = 22,
  Underline = 4,
  UnderlineOff = 24,
  Blink = 5,
  BlinkOff = 25,
  InvertColor = 7,
  InvertColorOff = 27,
  Hidden = 8,
  HiddenOff = 28
}

/**
 * The static functions on this class are used to produce colored text
 * for use with a terminal that supports ANSI escape codes.
 *
 * Note that this API always generates color codes, regardless of whether
 * the process's stdout is a TTY. The reason is that, in a complex program, the
 * code that is generating strings often does not know were those strings will end
 * up. In some cases, the same log message may get printed both to a shell
 * that supports color AND to a log file that does not.
 *
 * @example
 * ```ts
 * console.log(Colorize.red('Red Text!'))
 * terminal.writeLine(Colorize.green('Green Text!'), ' ', Colorize.blue('Blue Text!'));
 *```
 *
 * @public
 */
export class Colorize {
  public static black(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.BlackForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static red(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.RedForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static green(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.GreenForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static yellow(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.YellowForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static blue(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.BlueForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static magenta(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.MagentaForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static cyan(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.CyanForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static white(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.WhiteForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static gray(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.GrayForeground,
      SgrParameterAttribute.DefaultForeground,
      text
    );
  }

  public static blackBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.BlackBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static redBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.RedBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static greenBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.GreenBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static yellowBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.YellowBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static blueBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.BlueBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static magentaBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.MagentaBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static cyanBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.CyanBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static whiteBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.WhiteBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static grayBackground(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.GrayBackground,
      SgrParameterAttribute.DefaultBackground,
      text
    );
  }

  public static bold(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.Bold,
      SgrParameterAttribute.NormalColorOrIntensity,
      text
    );
  }

  public static dim(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.Dim,
      SgrParameterAttribute.NormalColorOrIntensity,
      text
    );
  }

  public static underline(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.Underline,
      SgrParameterAttribute.UnderlineOff,
      text
    );
  }

  public static blink(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.Blink,
      SgrParameterAttribute.BlinkOff,
      text
    );
  }

  public static invertColor(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.InvertColor,
      SgrParameterAttribute.InvertColorOff,
      text
    );
  }

  public static hidden(text: string): string {
    return Colorize._wrapTextInAnsiEscapeCodes(
      SgrParameterAttribute.Hidden,
      SgrParameterAttribute.HiddenOff,
      text
    );
  }

  private static _wrapTextInAnsiEscapeCodes(startCode: number, endCode: number, text: string): string {
    return (
      AnsiEscape.getEscapeSequenceForAnsiCode(startCode) +
      text +
      AnsiEscape.getEscapeSequenceForAnsiCode(endCode)
    );
  }
}
