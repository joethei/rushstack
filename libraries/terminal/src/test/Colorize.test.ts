// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { createColorGrid } from './createColorGrid';
import { Colorize } from '../Colorize';
import { AnsiEscape } from '../AnsiEscape';

describe(Colorize.name, () => {
  test('writes color grid correctly', () => {
    let lineCount: number = 0;
    for (const line of createColorGrid()) {
      expect(line.map((linePart) => AnsiEscape.formatForTests(linePart))).toMatchSnapshot(
        `line ${lineCount++}`
      );
    }

    expect(lineCount).toMatchInlineSnapshot(`10`);
  });

  it('generates codes as expected', () => {
    type ColorsFunctionNames = {
      [K in keyof typeof Colorize]: typeof Colorize[K] extends (str: string) => string ? K : never;
    }[keyof typeof Colorize];
    function testColorFunction(functionName: ColorsFunctionNames): void {
      expect(Colorize[functionName]('x')).toMatchSnapshot(functionName);
    }

    testColorFunction('black');
    testColorFunction('red');
    testColorFunction('green');
    testColorFunction('yellow');
    testColorFunction('blue');
    testColorFunction('magenta');
    testColorFunction('cyan');
    testColorFunction('white');
    testColorFunction('gray');
    testColorFunction('blackBackground');
    testColorFunction('redBackground');
    testColorFunction('greenBackground');
    testColorFunction('yellowBackground');
    testColorFunction('blueBackground');
    testColorFunction('magentaBackground');
    testColorFunction('cyanBackground');
    testColorFunction('whiteBackground');
    testColorFunction('grayBackground');
    testColorFunction('bold');
    testColorFunction('dim');
    testColorFunction('underline');
    testColorFunction('blink');
    testColorFunction('invertColor');
    testColorFunction('hidden');
  });
});
