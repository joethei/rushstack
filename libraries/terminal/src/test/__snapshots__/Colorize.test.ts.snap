// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Colorize generates codes as expected: black 1`] = `"[30mx[39m"`;

exports[`Colorize generates codes as expected: blackBackground 1`] = `"[40mx[49m"`;

exports[`Colorize generates codes as expected: blink 1`] = `"[5mx[25m"`;

exports[`Colorize generates codes as expected: blue 1`] = `"[34mx[39m"`;

exports[`Colorize generates codes as expected: blueBackground 1`] = `"[44mx[49m"`;

exports[`Colorize generates codes as expected: bold 1`] = `"[1mx[22m"`;

exports[`Colorize generates codes as expected: cyan 1`] = `"[36mx[39m"`;

exports[`Colorize generates codes as expected: cyanBackground 1`] = `"[46mx[49m"`;

exports[`Colorize generates codes as expected: dim 1`] = `"[2mx[22m"`;

exports[`Colorize generates codes as expected: gray 1`] = `"[90mx[39m"`;

exports[`Colorize generates codes as expected: grayBackground 1`] = `"[100mx[49m"`;

exports[`Colorize generates codes as expected: green 1`] = `"[32mx[39m"`;

exports[`Colorize generates codes as expected: greenBackground 1`] = `"[42mx[49m"`;

exports[`Colorize generates codes as expected: hidden 1`] = `"[8mx[28m"`;

exports[`Colorize generates codes as expected: invertColor 1`] = `"[7mx[27m"`;

exports[`Colorize generates codes as expected: magenta 1`] = `"[35mx[39m"`;

exports[`Colorize generates codes as expected: magentaBackground 1`] = `"[45mx[49m"`;

exports[`Colorize generates codes as expected: red 1`] = `"[31mx[39m"`;

exports[`Colorize generates codes as expected: redBackground 1`] = `"[41mx[49m"`;

exports[`Colorize generates codes as expected: underline 1`] = `"[4mx[24m"`;

exports[`Colorize generates codes as expected: white 1`] = `"[37mx[39m"`;

exports[`Colorize generates codes as expected: whiteBackground 1`] = `"[47mx[49m"`;

exports[`Colorize generates codes as expected: yellow 1`] = `"[33mx[39m"`;

exports[`Colorize generates codes as expected: yellowBackground 1`] = `"[43mx[49m"`;

exports[`Colorize writes color grid correctly: line 0 1`] = `
Array [
  "X",
  "[black]X[default]",
  "[white]X[default]",
  "[gray]X[default]",
  "[magenta]X[default]",
  "[red]X[default]",
  "[yellow]X[default]",
  "[green]X[default]",
  "[cyan]X[default]",
  "[blue]X[default]",
]
`;

exports[`Colorize writes color grid correctly: line 1 1`] = `
Array [
  "[black-bg]X[default-bg]",
  "[black-bg][black]X[default][default-bg]",
  "[black-bg][white]X[default][default-bg]",
  "[black-bg][gray]X[default][default-bg]",
  "[black-bg][magenta]X[default][default-bg]",
  "[black-bg][red]X[default][default-bg]",
  "[black-bg][yellow]X[default][default-bg]",
  "[black-bg][green]X[default][default-bg]",
  "[black-bg][cyan]X[default][default-bg]",
  "[black-bg][blue]X[default][default-bg]",
]
`;

exports[`Colorize writes color grid correctly: line 2 1`] = `
Array [
  "[white-bg]X[default-bg]",
  "[white-bg][black]X[default][default-bg]",
  "[white-bg][white]X[default][default-bg]",
  "[white-bg][gray]X[default][default-bg]",
  "[white-bg][magenta]X[default][default-bg]",
  "[white-bg][red]X[default][default-bg]",
  "[white-bg][yellow]X[default][default-bg]",
  "[white-bg][green]X[default][default-bg]",
  "[white-bg][cyan]X[default][default-bg]",
  "[white-bg][blue]X[default][default-bg]",
]
`;

exports[`Colorize writes color grid correctly: line 3 1`] = `
Array [
  "[gray-bg]X[default-bg]",
  "[gray-bg][black]X[default][default-bg]",
  "[gray-bg][white]X[default][default-bg]",
  "[gray-bg][gray]X[default][default-bg]",
  "[gray-bg][magenta]X[default][default-bg]",
  "[gray-bg][red]X[default][default-bg]",
  "[gray-bg][yellow]X[default][default-bg]",
  "[gray-bg][green]X[default][default-bg]",
  "[gray-bg][cyan]X[default][default-bg]",
  "[gray-bg][blue]X[default][default-bg]",
]
`;

exports[`Colorize writes color grid correctly: line 4 1`] = `
Array [
  "[magenta-bg]X[default-bg]",
  "[magenta-bg][black]X[default][default-bg]",
  "[magenta-bg][white]X[default][default-bg]",
  "[magenta-bg][gray]X[default][default-bg]",
  "[magenta-bg][magenta]X[default][default-bg]",
  "[magenta-bg][red]X[default][default-bg]",
  "[magenta-bg][yellow]X[default][default-bg]",
  "[magenta-bg][green]X[default][default-bg]",
  "[magenta-bg][cyan]X[default][default-bg]",
  "[magenta-bg][blue]X[default][default-bg]",
]
`;

exports[`Colorize writes color grid correctly: line 5 1`] = `
Array [
  "[red-bg]X[default-bg]",
  "[red-bg][black]X[default][default-bg]",
  "[red-bg][white]X[default][default-bg]",
  "[red-bg][gray]X[default][default-bg]",
  "[red-bg][magenta]X[default][default-bg]",
  "[red-bg][red]X[default][default-bg]",
  "[red-bg][yellow]X[default][default-bg]",
  "[red-bg][green]X[default][default-bg]",
  "[red-bg][cyan]X[default][default-bg]",
  "[red-bg][blue]X[default][default-bg]",
]
`;

exports[`Colorize writes color grid correctly: line 6 1`] = `
Array [
  "[yellow-bg]X[default-bg]",
  "[yellow-bg][black]X[default][default-bg]",
  "[yellow-bg][white]X[default][default-bg]",
  "[yellow-bg][gray]X[default][default-bg]",
  "[yellow-bg][magenta]X[default][default-bg]",
  "[yellow-bg][red]X[default][default-bg]",
  "[yellow-bg][yellow]X[default][default-bg]",
  "[yellow-bg][green]X[default][default-bg]",
  "[yellow-bg][cyan]X[default][default-bg]",
  "[yellow-bg][blue]X[default][default-bg]",
]
`;

exports[`Colorize writes color grid correctly: line 7 1`] = `
Array [
  "[green-bg]X[default-bg]",
  "[green-bg][black]X[default][default-bg]",
  "[green-bg][white]X[default][default-bg]",
  "[green-bg][gray]X[default][default-bg]",
  "[green-bg][magenta]X[default][default-bg]",
  "[green-bg][red]X[default][default-bg]",
  "[green-bg][yellow]X[default][default-bg]",
  "[green-bg][green]X[default][default-bg]",
  "[green-bg][cyan]X[default][default-bg]",
  "[green-bg][blue]X[default][default-bg]",
]
`;

exports[`Colorize writes color grid correctly: line 8 1`] = `
Array [
  "[cyan-bg]X[default-bg]",
  "[cyan-bg][black]X[default][default-bg]",
  "[cyan-bg][white]X[default][default-bg]",
  "[cyan-bg][gray]X[default][default-bg]",
  "[cyan-bg][magenta]X[default][default-bg]",
  "[cyan-bg][red]X[default][default-bg]",
  "[cyan-bg][yellow]X[default][default-bg]",
  "[cyan-bg][green]X[default][default-bg]",
  "[cyan-bg][cyan]X[default][default-bg]",
  "[cyan-bg][blue]X[default][default-bg]",
]
`;

exports[`Colorize writes color grid correctly: line 9 1`] = `
Array [
  "[blue-bg]X[default-bg]",
  "[blue-bg][black]X[default][default-bg]",
  "[blue-bg][white]X[default][default-bg]",
  "[blue-bg][gray]X[default][default-bg]",
  "[blue-bg][magenta]X[default][default-bg]",
  "[blue-bg][red]X[default][default-bg]",
  "[blue-bg][yellow]X[default][default-bg]",
  "[blue-bg][green]X[default][default-bg]",
  "[blue-bg][cyan]X[default][default-bg]",
  "[blue-bg][blue]X[default][default-bg]",
]
`;
